contracts/Rewarder.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.2;
+
+import "./Energy.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "hardhat/console.sol";
+
+contract Rewarder is Ownable {
+  Energy public energy;
+
+  struct Objective {
+    uint256 startingBlock;
+    uint256 endingBlock;
+    uint256 reward;
+    address origin;
+  }
+
+  Objective[] private objectives;
+
+  // origin address => user address => block list
+  mapping(address => mapping(address => uint256[])) public userInteractionsByOrigin;
+  // user address => objectiveIndex => claimed
+  mapping(address => mapping(uint256 => bool)) public objectivesCompletedByUser;
+
+  // un nombre mejor?
+  event UserInteraction(address indexed user, uint256 interactionIndex, address indexed originAddress);
+  event ObjectiveAdded(uint256 indexed index, uint256 reward, address indexed creator);
+  event RewardClaimed(address indexed user, uint256 reward, uint256 indexed objectiveIndex);
+
+  modifier hasObjective(address origin) {
+    bool objective = false;
+    for (uint256 i = 0; i < objectives.length; i++) {
+      if (objectives[i].origin == origin) {
+        objective = true;
+        break;
+      }
+    }
+    require(objective, "No reward assigned to msg.sender");
+    _;
+  }
+
+  constructor(address _energy) {
+    energy = Energy(_energy);
+  }
+
+  function getObjectives() external view returns (Objective[] memory) {
+    return objectives;
+  }
+
+  function onReward(address user) public hasObjective(msg.sender) returns (uint256 interactionIndex) {
+    userInteractionsByOrigin[msg.sender][user].push(block.number);
+    interactionIndex = userInteractionsByOrigin[msg.sender][user].length - 1;
+    emit UserInteraction(user, interactionIndex, msg.sender);
+    return interactionIndex;
+  }
+
+  function getObjectiveByOrigin(address origin) external view returns (int256 index) {
```

Comment:
Same problem about iterating through all objetives.

In this case though I see that it's only used by a test, so maybe we can just add a warning not to use this on-chain, just for off-line computations.

Another option is to have the objetives be a mapping from origin to list of objectives instead of a plain list. Might give us more flexibility for the future.

I think we also need a `getObjectiveByIndex` function. That way we can figure out which objetive we care about off-chain and just pass the index on-chain to access in O(1). Iterating through a dynamic array on-chain is cost prohivitive :/

Reply:
I think the last option (`getObjectiveByIndex` function) is probably the best. In any case I'm not sure why we'd need to call either of these functions, but it's good to have utility functions to make our lives easier.

The "optional" functions such as these I'd put at the bottom of the file with a comment about it

Reply:
great! I'll just keep `getObjectiveByIndex`. I think it could be useful if we are claiming the reward directly from the contract that called the rewarder (it should have a flag argument like `shouldClaim`), but not sure yet.

Reply:
Right, yeah that makes sense